agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentWebSocketClient.kt

```
@@ -0,0 +1,211 @@
+
+    override fun onClose(code: Int, reason: String?, remote: Boolean) {
+        log.info { "$socketName:AgentConnection closed: $code,$reason,$remote" }
+    }
+
+    /**
+     * inbound messages list
+     */
+    private val receivedMessages = ConcurrentHashMap<String, ConcurrentLinkedDeque<String>>()
+
+    /**
+     * message routing map
+     */
+    private val subscribedObservers = ConcurrentHashMap<String, ConcurrentLinkedDeque<SingleSubscriber<in String>>>()
+
+    /**
+     * Adds an observer to (FIFO) queue corresponding to the given key.
+     * If the queue doesn't exist for the given key, it's been created.
+     */
+    private fun addObserver(key: String, observer: SingleSubscriber<in String>) =
+            subscribedObservers.getOrPut(key) { ConcurrentLinkedDeque() }.addLast(observer)
+
+    /**
+     * Adds a message to the queue corresponding to the given key.
+     * If the queue doesn't exist for the given key, it's been created.
+     */
+    private fun storeMessage(key: String, message: String) =
+            receivedMessages.getOrPut(key) { ConcurrentLinkedDeque() }.addLast(message)
+
+    /**
+     * Removes an observer from the queue.
+     */
+    private fun popObserver(key: String) =
+            subscribedObservers.getOrPut(key) { ConcurrentLinkedDeque() }.pollFirst()
+
+    /**
+     * Pops a message by key from the queue.
+     */
+    private fun popMessage(key: String) =
+            receivedMessages.getOrPut(key) { ConcurrentLinkedDeque() }.pollFirst()
+
+    /**
+     * Dispatches the message to an observer.
+     * Each message '@type' has own routing agreement
+     */
+    override fun onMessage(msg: String?) {
+        var message = msg
+        log.info { "$socketName:ReceivedMessage: $message" }
+        if (message != null) {
+            val obj = SerializationUtils.jSONToAny<ObjectNode>(message)
+            val type: String = obj["@type"].asText()
+            var key: String? = null
+            when (type) {
```

We have 2 types of "keys" for messages: type and composite key. Well, in which cases we store by type? I assume these types are not handled explicitly: MESSAGE_SENT, REQUEST_SENT, am I right? 

If yes, I would prefer to handle them explicitly, and fail if unknown type is received. In such case we will follow protocol changes and fail with proper message, otherwise it hides an error. 

`key = when(type) { /* TODO */ } if(key == null) throw 
`

---

We store by type everything except for those (4 message types) we explicitly specify the storage key for. We don't handle certain responses such as MESSAGE_SENT and RESPONSE_SENT as agreed with @adoroganov. This is to be done in the future versions.
Message specific "key" is either "type" or "type.pubkey", "type.DID", "class.DID" etc. It can't be null, so there's no use case for throwing an exception.